cuML/src/glm/qn/qn.h

```diff
@@ -42,61 +42,63 @@ int qn_fit(LossFunction &loss, T *Xptr, T *yptr, T *zptr, int N,
-           int loss_type, cublasHandle_t cublas, cudaStream_t stream) {
-
-  // TODO this will come from the cuml handle
+void qnFit(const cumlHandle_impl &cuml, T *X, T *y, int N, int D, int C,
+           bool fit_intercept, T l1, T l2, int max_iter, T grad_tol,
+           int linesearch_max_iter, int lbfgs_memory, int verbosity, T *w0,
+           T *f, int *num_iters, bool X_col_major, int loss_type,
+           cudaStream_t stream) {
 
   STORAGE_ORDER ord = X_col_major ? COL_MAJOR : ROW_MAJOR;
 
-  SimpleMatOwning<T> z(C, N); // TODO this allocation somewhere else?
+  MLCommon::device_buffer<T> tmp(cuml.getDeviceAllocator(), stream, C * N);
+
+  SimpleMat<T> z(tmp.data(), C, N);
 
   switch (loss_type) {
   case 0: {
     ASSERT(C == 1, "qn.h: logistic loss invalid C");
-    LogisticLoss<T> loss(D, fit_intercept, cublas);
-    qn_fit<T, decltype(loss)>(loss, X, y, z.data, N, fit_intercept, l1, l2,
-                               max_iter, grad_tol, linesearch_max_iter,
-                               lbfgs_memory, verbosity, w0, f, num_iters, ord,
-                               stream);
+    LogisticLoss<T> loss(D, fit_intercept, cuml);
+    qn_fit<T, decltype(loss)>(cuml, loss, X, y, z.data, N, fit_intercept, l1, l2,
+                              max_iter, grad_tol, linesearch_max_iter,
+                              lbfgs_memory, verbosity, w0, f, num_iters, ord,
+                              stream);
   } break;
   case 1: {
 
     ASSERT(C == 1, "qn.h: squared loss invalid C");
-    SquaredLoss<T> loss(D, fit_intercept, cublas);
-    qn_fit<T, decltype(loss)>(loss, X, y, z.data, N, fit_intercept, l1, l2,
-                               max_iter, grad_tol, linesearch_max_iter,
-                               lbfgs_memory, verbosity, w0, f, num_iters, ord,
-                               stream);
+    SquaredLoss<T> loss(D, fit_intercept, cuml);
+    qn_fit<T, decltype(loss)>(cuml, loss, X, y, z.data, N, fit_intercept, l1, l2,
+                              max_iter, grad_tol, linesearch_max_iter,
+                              lbfgs_memory, verbosity, w0, f, num_iters, ord,
+                              stream);
   } break;
   case 2: {
 
     ASSERT(C > 1, "qn.h: softmax invalid C");
-    Softmax<T> loss(D, C, fit_intercept, cublas);
-    qn_fit<T, decltype(loss)>(loss, X, y, z.data, N, fit_intercept, l1, l2,
-                               max_iter, grad_tol, linesearch_max_iter,
-                               lbfgs_memory, verbosity, w0, f, num_iters, ord,
-                               stream);
+    Softmax<T> loss(D, C, fit_intercept, cuml);
+    qn_fit<T, decltype(loss)>(cuml, loss, X, y, z.data, N, fit_intercept, l1, l2,
+                              max_iter, grad_tol, linesearch_max_iter,
+                              lbfgs_memory, verbosity, w0, f, num_iters, ord,
+                              stream);
   } break;
   default: { ASSERT(false, "qn.h: unknown loss function."); }
   }
+  tmp.release(stream);
```

Comment:
This is done implicitly when `tmp` get's out of scope. Any reason to do this explicitly? Also applies to other files.